ksql-examples/src/main/java/io/confluent/ksql/datagen/DataGenProducer.java

```diff
@@ -320,4 +357,70 @@ private String getRandomToken(Set<String> collected) {
       org.apache.kafka.connect.data.Schema kafkaSchema,
       String topicName
   );
+
+  private org.apache.kafka.connect.data.Schema getOptionalSchema(
+      final org.apache.kafka.connect.data.Schema schema) {
+    switch (schema.type()) {
+      case BOOLEAN:
+        return org.apache.kafka.connect.data.Schema.OPTIONAL_BOOLEAN_SCHEMA;
+      case INT32:
+        return org.apache.kafka.connect.data.Schema.OPTIONAL_INT32_SCHEMA;
+      case INT64:
+        return org.apache.kafka.connect.data.Schema.OPTIONAL_INT64_SCHEMA;
+      case FLOAT64:
+        return org.apache.kafka.connect.data.Schema.OPTIONAL_FLOAT64_SCHEMA;
+      case STRING:
+        return org.apache.kafka.connect.data.Schema.OPTIONAL_STRING_SCHEMA;
+      case ARRAY:
+        return SchemaBuilder.array(getOptionalSchema(schema.valueSchema())).optional().build();
+      case MAP:
+        return SchemaBuilder.map(
+            getOptionalSchema(schema.keySchema()),
+            getOptionalSchema(schema.valueSchema()))
+            .optional().build();
+      case STRUCT:
+        final SchemaBuilder schemaBuilder = SchemaBuilder.struct();
+        for (Field field : schema.fields()) {
+          schemaBuilder.field(field.name(), getOptionalSchema(field.schema()));
+        }
+        return schemaBuilder.optional().build();
+      default:
+        throw new KsqlException("Unsupported type: " + schema);
+    }
+  }
+
+  private Object getOptionalValue(
+      final org.apache.kafka.connect.data.Schema schema,
+      final Object value) {
+    switch (schema.type()) {
+      case BOOLEAN:
+      case INT32:
+      case INT64:
+      case FLOAT64:
+      case STRING:
+        return value;
+      case ARRAY:
+        final List<?> list = (List<?>) value;
+        return list.stream().map(listItem -> getOptionalValue(schema.valueSchema(), listItem))
+            .collect(Collectors.toList());
+      case MAP:
+        final Map<?, ?> map = (Map<?, ?>) value;
+        return map.entrySet().stream()
+            .collect(Collectors.toMap(
+                k -> getOptionalValue(schema.keySchema(), k),
+                v -> getOptionalValue(schema.valueSchema(), v)
+            ));
+      case STRUCT:
+        final Struct struct = (Struct) value;
+        final Struct optionalStruct = new Struct(getOptionalSchema(schema));
```

Comment:
For my understanding, it seems like that the sole purpose of this method is to generate an optional schema for a struct value?

Reply:
Correct.